module/zfs/dbuf.c

```diff
@@ -4574,12 +4585,22 @@ dbuf_lightweight_ready(zio_t *zio)
 {
 	dbuf_dirty_record_t *dr = zio->io_private;
 	blkptr_t *bp = zio->io_bp;
+	dmu_buf_impl_t *parent_db = NULL;
 
 	if (zio->io_error != 0)
 		return;
 
 	dnode_t *dn = dr->dr_dnode;
 
+	EQUIV(dr->dr_parent == NULL, dn->dn_phys->dn_nlevels == 1);
+	if (dr->dr_parent == NULL) {
+		parent_db = dn->dn_dbuf;
+	} else {
+		parent_db = dr->dr_parent->dr_dbuf;
+	}
+	mutex_enter(&parent_db->db_mtx);
+
+	rw_enter(&parent_db->db_rwlock, RW_READER);
```

Comment:
OK, this we might need for purposes of `bp_get_dsize_sync()` and later.  But since we are going to upgrade it any way, is there a reason to not to RW_WRITER from the beginning?